kratos/solving_strategies/builder_and_solvers/p_multigrid/constraint_utilities.hpp

```diff
@@ -0,0 +1,372 @@
+                                         [[maybe_unused]] MasterSlaveConstraint::VectorType& rConstraintGaps,
+                                         const MasterSlaveConstraint& rConstraint,
+                                         const std::vector<std::size_t>& rSlaveIds,
+                                         const std::vector<std::size_t>& rMasterIds,
+                                         const std::unordered_map<std::size_t,std::pair<std::size_t,std::size_t>>& rConstraintIdMap)
+{
+    // Constraint identifiers are the slave DoFs' IDs.
+    rConstraintIndices.resize(rSlaveIds.size());
+    std::transform(rSlaveIds.begin(),
+                   rSlaveIds.end(),
+                   rConstraintIndices.begin(),
+                   [&rConstraintIdMap](std::size_t slave_id){
+                        return rConstraintIdMap.at(slave_id).first;
+                   });
+
+    // Comb DoFs.
+    rDofIds.resize(rMasterIds.size() + rSlaveIds.size());
+    std::copy(rMasterIds.begin(), rMasterIds.end(), rDofIds.begin());
+    std::copy(rSlaveIds.begin(), rSlaveIds.end(), rDofIds.begin() + rMasterIds.size());
+
+    // Reconstruct the constraint equations.
+    if (rRelationMatrix.size2()) {
+        rRelationMatrix.resize(rRelationMatrix.size1(),
+                               rRelationMatrix.size2() + rSlaveIds.size(),
+                               /*preserve entries=*/true);
+        for (std::size_t i_slave=0ul; i_slave<rSlaveIds.size(); ++i_slave) {
+            // Fetch the number of constraint objects defining this constraint equation,
+            // because the slave DoF's coefficient must add up to -1 after assembly.
+            const auto constraint_object_count = rConstraintIdMap.at(rSlaveIds[i_slave]).second;
+            rRelationMatrix(i_slave, rMasterIds.size() + i_slave) = -1.0 / constraint_object_count;
+        } // for i_slave in range(rSlaveIds.size())
+    } // if rRelationMatrix.size2()
+
+    // No need to move the constraint gaps from RHS to LHS
+    // because the slaves were moved to the RHS instead:
+    // u^s = T u^m + b
+    // =>
+    // 0 = A [u^m \\ u^s] + b
+    //std::transform(rConstraintGaps.begin(),
+    //               rConstraintGaps.end(),
+    //               rConstraintGaps.begin(),
+    //               [](auto constraint_gap){return -constraint_gap;});
+}
+
+
+inline void ProcessMultifreedomConstraint(std::vector<std::size_t>& rConstraintIndices,
+                                          std::vector<std::size_t>& rDofIds,
+                                          const MasterSlaveConstraint& rConstraint,
+                                          const std::vector<std::size_t>& rMasterIds,
+                                          const std::unordered_map<std::size_t,std::pair<std::size_t,std::size_t>>& rConstraintIdMap)
+{
+    const auto& r_constraint_labels = rConstraint.GetData().GetValue(CONSTRAINT_LABELS);
```

Comment:
What are the `CONSTRAINT_LABELS`? A description would be nice.

Reply:
it's not a variable I added so I can't document its purpose, but I can add docs about what purpose it serves within the `ConstraintAssembler` system.

In fact, I already did just that. Take a look at the docs of `MultifreedomConstraint`.